src/rgw/rgw_cache.h

```diff
@@ -216,4 +361,219 @@ class ObjectCache {
   void invalidate_all();
 };
 
+struct get_obj_data;
+
+template <class T>
+class RGWDataCache : public T
+{
+
+  DataCache   data_cache;
+
+public:
+  RGWDataCache() {}
+
+  int init_rados() override {
+    int ret;
+    data_cache.init(T::cct);
+    ret = T::init_rados();
+    if (ret < 0)
+      return ret;
+
+    return 0;
+  }
+
+  int flush_read_list(struct get_obj_data* d);
+  int get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                         off_t read_ofs, off_t len, bool is_head_obj,
+                         RGWObjState *astate, void *arg) override;
+};
+
+
+template<typename T>
+int RGWDataCache<T>::flush_read_list(struct get_obj_data* d) {
+
+  d->data_lock.lock();
+  std::list<bufferlist> l;
+  l.swap(d->read_list);
+  d->get();
+  d->read_list.clear();
+  d->data_lock.unlock();
+
+  int r = 0;
+
+  std::string oid;
+  std::list<bufferlist>::iterator iter;
+  for (iter = l.begin(); iter != l.end(); ++iter) {
+    bufferlist& bl = *iter;
+    oid = d->get_pending_oid();
+    if (bl.length() == 0x400000)
+      data_cache.put(bl, bl.length(), oid);
```

Comment:
`get_pending_oid()` can return an empty string; don't we need to handle that case here?

Reply:
adding a check and log + error handling
(checked a test run with normal workload exercising the cache, empty oid was not returned)